autoload/ale/floating_preview.vim

```diff
@@ -170,7 +172,14 @@ function! s:VimCreate(options) abort
     \        get(g:ale_floating_window_border, 5, '+'),
     \    ],
     \    'moved': 'any',
-    \    })
+    \    }
+
+    " allow custom popup opts
+    if exists('g:ale_floating_preview_popup_opts')
```

Comment:
I would allow this to be either a function name or a dictionary. Then we can check do something like:

```vimscript
if exists('g:ale_floating_preview_popup_opts')
    if type(g:ale_floating_preview_popup_opts) is# v:t_dict
        let l:popup_opts = extend(l:popup_opts, g:ale_floating_preview_popup_opts)
    elseif exists('*' . g:ale_floating_preview_popup_opts)
        let l:popup_opts = extend(l:popup_opts, function(g:ale_floating_preview_popup_opts)())
    endif
endif
```

Note that:

  1. I am extending the defaults instead of replacing them. This allows users to set some of the options without need to redefine them all.
  2. I check if the function does exists before invoking it.

Reply:
Allowing for either a dictionary or a function is a great idea, I'll add that. Originally I wrote this using extend like you have suggested, but I wanted to be able to specify only the options that I wanted. I think this is more specific to my usage, so extending the default options is probably a better idea for most users.

On checking if the funcref actually exists, in my setup I am referencing an autoload function that fails the exists() check because it hasn't been loaded yet. Example:

```vim
let g:ale_floating_preview_popup_opts = 'myplugin#get_popup_opts'
```

Do you have a suggestion on how best to handle this? I'm happy to adjust my setup if you prefer to keep the exists check.

Reply:
Nice, I tested this and works nicely in Neovim with a dictionary and a funcref.  

The idea of checking the funcref is that if the user mistakenly sets something wrong we don't get ALE spitting cryptic messages.  Something like this would do and even cover other cases like setting a number instead of a string or funcref:

```vimscript
function! s:GetPopupOpts() abort
    if !exists('g:ale_floating_preview_popup_opts')
        return {}
    endif

    let l:ref = g:ale_floating_preview_popup_opts

    if type(l:ref) is# v:t_dict
        return l:ref
    endif

    if type(l:ref) is# v:t_string
        try
            return function(l:ref)()
        catch E700
        endtry
    endif

    return {}
endfunction
```

Reply:
Perfect! I've modified this now to try/catch if it's a string.